File: mcp/atlas-mcp-autopilot/src/receipts.ts
/**
 * Receipts Module - Audit trail for all tool executions.
 */

import { v4 as uuidv4 } from 'uuid';
import { getDb } from '../db/index.js';
import type { PolicyDecision } from '../policy/index.js';

export interface Receipt {
  receipt_id: string;
  timestamp_utc: string;
  tool_name: string;
  args_json: string;
  decision: PolicyDecision;
  result: 'OK' | 'ERROR' | 'PENDING_CONFIRM';
  changes_json: string;
  undo_json: string;
  stdout?: string;
  stderr?: string;
  exit_code?: number;
  error_json?: string;
  undo_supported: boolean;
}

export interface ReceiptInput {
  tool_name: string;
  args: Record<string, unknown>;
  decision: PolicyDecision;
  result: 'OK' | 'ERROR' | 'PENDING_CONFIRM';
  changes?: unknown[];
  undo?: unknown[];
  stdout?: string;
  stderr?: string;
  exit_code?: number;
  error?: unknown;
  undo_supported?: boolean;
}

/**
 * Create and store a new receipt.
 */
export function createReceipt(input: ReceiptInput): Receipt {
  const receipt: Receipt = {
    receipt_id: uuidv4(),
    timestamp_utc: new Date().toISOString(),
    tool_name: input.tool_name,
    args_json: JSON.stringify(input.args),
    decision: input.decision,
    result: input.result,
    changes_json: JSON.stringify(input.changes || []),
    undo_json: JSON.stringify(input.undo || []),
    stdout: input.stdout,
    stderr: input.stderr,
    exit_code: input.exit_code,
    error_json: input.error ? JSON.stringify(input.error) : undefined,
    undo_supported: input.undo_supported ?? false,
  };
  
  const db = getDb();
  db.prepare(`
    INSERT INTO receipts (
      receipt_id, timestamp_utc, tool_name, args_json, decision, result,
      changes_json, undo_json, stdout, stderr, exit_code, error_json, undo_supported
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `).run(
    receipt.receipt_id,
    receipt.timestamp_utc,
    receipt.tool_name,
    receipt.args_json,
    receipt.decision,
    receipt.result,
    receipt.changes_json,
    receipt.undo_json,
    receipt.stdout || null,
    receipt.stderr || null,
    receipt.exit_code ?? null,
    receipt.error_json || null,
    receipt.undo_supported ? 1 : 0
  );
  
  return receipt;
}

/**
 * List receipts with optional limit.
 */
export function listReceipts(limit: number = 50): Receipt[] {
  const db = getDb();
  const rows = db.prepare(`
    SELECT * FROM receipts ORDER BY timestamp_utc DESC LIMIT ?
  `).all(limit) as Array<{
    receipt_id: string;
    timestamp_utc: string;
    tool_name: string;
    args_json: string;
    decision: PolicyDecision;
    result: 'OK' | 'ERROR' | 'PENDING_CONFIRM';
    changes_json: string;
    undo_json: string;
    stdout: string | null;
    stderr: string | null;
    exit_code: number | null;
    error_json: string | null;
    undo_supported: number;
  }>;
  
  return rows.map(row => ({
    ...row,
    stdout: row.stdout || undefined,
    stderr: row.stderr || undefined,
    exit_code: row.exit_code ?? undefined,
    error_json: row.error_json || undefined,
    undo_supported: Boolean(row.undo_supported),
  }));
}

/**
 * Get a receipt by ID.
 */
export function getReceipt(receiptId: string): Receipt | null {
  const db = getDb();
  const row = db.prepare(`
    SELECT * FROM receipts WHERE receipt_id = ?
  `).get(receiptId) as {
    receipt_id: string;
    timestamp_utc: string;
    tool_name: string;
    args_json: string;
    decision: PolicyDecision;
    result: 'OK' | 'ERROR' | 'PENDING_CONFIRM';
    changes_json: string;
    undo_json: string;
    stdout: string | null;
    stderr: string | null;
    exit_code: number | null;
    error_json: string | null;
    undo_supported: number;
  } | undefined;
  
  if (!row) return null;
  
  return {
    ...row,
    stdout: row.stdout || undefined,
    stderr: row.stderr || undefined,
    exit_code: row.exit_code ?? undefined,
    error_json: row.error_json || undefined,
    undo_supported: Boolean(row.undo_supported),
  };
}

File: apps/api/src/atlas/storage/receipts.py
"""
Receipts Store - Persistence layer for execution receipts.

Every execution in ATLAS produces a receipt. This store ensures
receipts are persisted reliably and can be queried for debugging,
auditing, and undo operations.
"""

import json
from datetime import datetime
from typing import Any
from uuid import UUID

from atlas.core.models import Receipt, ReceiptStatus
from atlas.storage.database import Database


class ReceiptsStore:
    """
    Store for execution receipts.
    
    Provides CRUD operations for receipts with full JSON storage
    for complete audit trail.
    """

    def __init__(self, database: Database):
        self.db = database

    async def create(self, receipt: Receipt) -> Receipt:
        """
        Persist a new receipt.
        
        Args:
            receipt: The receipt to store
            
        Returns:
            The stored receipt (unchanged)
        """
        receipt_json = receipt.model_dump_json()
        
        await self.db.execute(
            """
            INSERT INTO receipts (receipt_id, status, user_input, receipt_json)
            VALUES (?, ?, ?, ?)
            """,
            (
                str(receipt.receipt_id),
                receipt.status.value,
                receipt.user_input,
                receipt_json,
            ),
        )
        
        # Commit the transaction
        if self.db._connection:
            await self.db._connection.commit()
        
        return receipt

    async def get(self, receipt_id: str | UUID) -> Receipt | None:
        """
        Get a receipt by ID.
        
        Args:
            receipt_id: The receipt UUID (string or UUID)
            
        Returns:
            The receipt or None if not found
        """
        row = await self.db.fetch_one(
            "SELECT receipt_json FROM receipts WHERE receipt_id = ?",
            (str(receipt_id),),
        )
        
        if row:
            return Receipt.model_validate_json(row["receipt_json"])
        return None

    async def list(
        self,
        limit: int = 50,
        offset: int = 0,
        status: ReceiptStatus | None = None,
    ) -> list[Receipt]:
        """
        List receipts with optional filtering.
        
        Args:
            limit: Maximum number of receipts to return
            offset: Number of receipts to skip
            status: Optional status filter
            
        Returns:
            List of receipts, newest first
        """
        if status:
            rows = await self.db.fetch_all(
                """
                SELECT receipt_json FROM receipts 
                WHERE status = ?
                ORDER BY created_at DESC 
                LIMIT ? OFFSET ?
                """,
                (status.value, limit, offset),
            )
        else:
            rows = await self.db.fetch_all(
                """
                SELECT receipt_json FROM receipts 
                ORDER BY created_at DESC 
                LIMIT ? OFFSET ?
                """,
                (limit, offset),
            )
        
        return [Receipt.model_validate_json(row["receipt_json"]) for row in rows]

    async def count(self, status: ReceiptStatus | None = None) -> int:
        """
        Count total receipts.
        
        Args:
            status: Optional status filter
            
        Returns:
            Total count
        """
        if status:
            row = await self.db.fetch_one(
                "SELECT COUNT(*) as count FROM receipts WHERE status = ?",
                (status.value,),
            )
        else:
            row = await self.db.fetch_one("SELECT COUNT(*) as count FROM receipts")
        
        return row["count"] if row else 0

    async def update(self, receipt: Receipt) -> Receipt:
        """
        Update an existing receipt.
        
        Used when confirming pending tool calls or updating status.
        
        Args:
            receipt: The updated receipt
            
        Returns:
            The updated receipt
        """
        receipt_json = receipt.model_dump_json()
        
        await self.db.execute(
            """
            UPDATE receipts 
            SET status = ?, receipt_json = ?
            WHERE receipt_id = ?
            """,
            (receipt.status.value, receipt_json, str(receipt.receipt_id)),
        )
        
        if self.db._connection:
            await self.db._connection.commit()
        
        return receipt

    async def delete(self, receipt_id: str | UUID) -> bool:
        """
        Delete a receipt by ID.
        
        Args:
            receipt_id: The receipt UUID
            
        Returns:
            True if deleted, False if not found
        """
        cursor = await self.db.execute(
            "DELETE FROM receipts WHERE receipt_id = ?",
            (str(receipt_id),),
        )
        
        if self.db._connection:
            await self.db._connection.commit()
        
        return cursor.rowcount > 0

    async def get_recent(self, hours: int = 24, limit: int = 100) -> list[Receipt]:
        """
        Get recent receipts within a time window.
        
        Args:
            hours: Number of hours to look back
            limit: Maximum receipts to return
            
        Returns:
            List of recent receipts
        """
        rows = await self.db.fetch_all(
            """
            SELECT receipt_json FROM receipts 
            WHERE created_at >= datetime('now', ?)
            ORDER BY created_at DESC 
            LIMIT ?
            """,
            (f"-{hours} hours", limit),
        )
        
        return [Receipt.model_validate_json(row["receipt_json"]) for row in rows]

    async def get_by_status(
        self, status: ReceiptStatus, limit: int = 50
    ) -> list[Receipt]:
        """
        Get receipts by status.
        
        Useful for finding pending confirmations or failed executions.
        
        Args:
            status: The status to filter by
            limit: Maximum receipts to return
            
        Returns:
            List of matching receipts
        """
        return await self.list(limit=limit, status=status)
